### ISSUE TITLE
AcceptSession and ReceiveAsync tasks are still running after QueueClient is closed

### ISSUE THREAD

**Issue by lmolkova** [Azure/azure-service-bus-dotnet#372]
## Actual Behavior

```C#
           queueClient.RegisterSessionHandler(
                (session, msg, ct) => Task.CompletedTask;
                exArgs => Task.CompletedTask);
           await queueClient.CloseAsync();
```
When queueClient is being closed it closes SessionPumpHost.
SessionPumpHost cancels `sessionPumpCancellationTokenSource` that should stop all running tasks, however 
in `SessionReceivePump.SessionPumpTaskAsync` `AcceptMessageSessionAsync` is called without cancellationToken  and  in absence of any session to accept, will terminate with SBTimeoutException after OperationTimeout and all the retries.

```C#
var session = await this.client.AcceptMessageSessionAsync().ConfigureAwait(false);
```

Similar behavior should be observed with `MessageReceivePump.MessagePumpTaskAsync` as `ReceiveAsync` also does not support cancellationTokens

```C#
message = await this.messageReceiver.ReceiveAsync(this.registerHandlerOptions.ReceiveTimeOut).ConfigureAwait(false);
```
## Expected Behavior
After QueueClient (MessageReceivePump/SessionReceivePump) are closed, all underlying tasks are temrinated.

## Versions  
- OS platform and version: any
- .NET Version: any
- NuGet package version or commit ID: 41561149a59fda88b6256866fb1f5067431292ab

**Comment by brutaldev** [comment]
I see something similar, the callback on `RegisterMessageHandler` continues to fire even when closing the receiver. The cancellation token is cancelled but message continue to stream in. Because the token is cancelled, nothing is processed which results in thousands of messages getting dead-lettered because of max delivery counts being exceeded.  How do you unregister/cancel the message handler when things are shutting down?

**Comment by nemakam** [comment]
@lmolkova 
Sorry for the late response.
I went through the code and don't see any anomaly regarding this. 

`SessionReceivePump` constructor takes in the `pumpCancellationToken` which is cancelled on `SessionPumpHost.Close`.

And each of the task inside the pump works within a `!IsCancelled` check in the while loop. (`SessionPumpAsyncTask` - `MessagePumpTaskAsync` - which then passes the cancellationToken to the user's callback).

Its true that neither `AcceptMessageSession` and `RecieveAsync` accepts cancellationToken. But none of the underneath tasks will get into the loop if `cancellationToken.IsCancelled` is `true`. 

@lmolkova  @brutaldev 
One thing to verify in your code is whether you check for `cancellationToken.IsCancelled` before processing your request? If the `ReceiveAsync()` was called before the cancellationToken was invoked, we will end up calling user-callback along with the cancellationToken. So the responsibility would be of the user's callback function.

**Comment by nemakam** [comment]
@brutaldev 
Unless there's a bug, max delivery count cannot be addressed here as we are talking about a scenario where `ReceiveAsync` was initiated before the token was cancelled. Since the cancellationToken token is client side token and not passed to service, whenever the message is received, the delivery count would have been updated on service side.
On a side note, consider increasing your MaxDeliveryCount setting if this is bothering your service.

**Comment by brutaldev** [comment]
@nemakam 
You can ignore the delivery count, with that I can see that the pump does not stop because eventually all messages end up dead-lettered with that reason. Inside the message callback handler, the first thing that is checked is the cancellation token so it skips all the handling logic. This causes the queue to drain rapidly. Using high concurrency and a short lock duration, the message return quite quickly and eventually hit their max delivery count. I've had queues with 10k messages drain AFTER closing the receiver.

I was looking at this all day yesterday and found that `CloseAsync` throws an exception so it doesn't appear to close the receiver, but does set the cancellation token to end up in this situation where receiver just dumps messages but none get processed. I can't see how this can so wrong in the `MessageReceivePump` by looking at the code, the same cancellation token stops the loop and is sent in the callback, unless it's changed a lot since Oct 12...

I've added a ton of `IsClosedOrClosing` checks everywhere now with additional logging so that I can try and catch this. When the services move around (Service Fabric) they get internally disposed which is when this happens. I'll feedback on whatever I find.

**Comment by nemakam** [comment]
@brutaldev 
What is the configuration of the MessagePump that you use?
How many concurrent requests? 
What is your max delivery count?
How long does your callback usually take to execute?

**Comment by nemakam** [comment]
@brutaldev 
Could you also enable tracing on `Microsoft-Azure-ServiceBus` source? I would like to see how many `ReceiveAsync` and `AcceptMessageSession` calls get triggered AFTER closing the receiver.

**Comment by lmolkova** [comment]
@nemakam 
I probably forgot some details, but as far as I remember, it's `AcceptSession` or `ReceiveAsync` that are waiting for timeout and then throw. So it easily repro on an empty queue. Just add handler and close the client than wait for timeout and see.

**Comment by nemakam** [comment]
@lmolkova 
Do you mean `exceptionReceivedHandler` is called after cancellationToken has been cancelled; and this need to be corrected?
Or even the `messageHandler` is invoked after closing the receiver?

**Comment by lmolkova** [comment]
I mean that if AcceptMessageSession was started and waits for the next message, it does not stop when queueClient is closed

`var session = await this.client.AcceptMessageSessionAsync().ConfigureAwait(false);`

https://github.com/Azure/azure-service-bus-dotnet/blob/e1bd055cf22b68eb43ac8ce2d10af8f7eb4e720e/src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs#L125

As you can see, it does not accept cancellation token, so closing queue client and all underlying objects has no effect on it. `AcceptMessageSession` task will continue to run until a message is received or the timeout happens.

**Comment by nemakam** [comment]
@lmolkova 
That's true. But once the session is accepted, it creates a new task where the first thing that is checked is ```
while (!this.pumpCancellationToken.IsCancellationRequested && !session.IsClosedOrClosing)
```
This is true for every task that is scheduled. 
So none of these would have negative repercussions (other than creation of the new task itself). No action is performed after the client has been closed.

**Comment by lmolkova** [comment]
If the session was accepted, then perhaps it would try to process it in `TaskExtensionHelper.Schedule(() => this.MessagePumpTaskAsync(messageSession));` as cancellation token is checked at the beginning of the loop. Perhaps underlying connections are closed by this time and `AcceptMessageSession` does not have a chance to receive something. However according to @brutaldev it could be the case.

Also, even if nothing is received AcceptMessageSession task continues to run that should not happen after close. As long as this task does not have a chance to receive something it's a subtle issue and should not affect anyone.

**Comment by nemakam** [comment]
There are multiple parts to this. As soon as `CloseAsync` is called, 
1. No more requests should go to server.
2. No more messages or sessions should come from server.
3. All tasks which are created on client should finish.

For 1st and 2nd point, the amqp-link to the server is shut down. Hence no new request reaches server. Similarly no new messages is received on the client.
Regarding third point, since cancellationToken is not immediate, it eventually shuts down/closes all the running tasks. 

If we can find any issue like @brutaldev 's case, we should definitely work on it. If its just presence of background tasks even after close, this might not be a problem that will warrant a lot of attention.

**Comment by lmolkova** [comment]
> If we can find any issue like @brutaldev 's case, we should definitely work on it. If its just presence of background tasks even after close, this might not be a problem that will warrant a lot of attention.

Agreed. Initially, the issue was created regarding p3 only and is pretty minor.

**Comment by vincehorst** [comment]
I've run into this issue in the real world. Do you have a recommendation on how to work around it?

**Comment by nemakam** [comment]
@vincehorst 
Could you describe more on how you use the client and how is this affecting your workload?

**Comment by SeanFeldman** [comment]
Related to #439

**Comment by vincehorst** [comment]
Hi guys, apologies but I am headed out of town for a couple of weeks vacation. Here is some quick untested pseudo code that describes the situation that I came across today. Feel free to ask follow up questions and I will respond when I get back.

```
// Connect to a queue that requires sessions.
// Ensure that the queue has no messages in it.
// Note the use of the TimeSpan for the serverWaitTime parameter on AcceptMessageSessionAsync() 
// Note that I am not checking whether there are existing sessions before trying to accept a session
for(var i = 0; i <= 2; i++){ 
	var queueClient = QueueClient.CreateFromConnectionString("<My Connection String>", ReceiveMode.PeekLock);
	try {
		var session = await queueClient.AcceptMessageSessionAsync("mySessionName", new System.TimeSpan(0, 0, 2));
	} catch (TimeoutException) {
		// On the first iteration a TimeoutException will be thrown if there are no sessions (messages) available. This is expected.
	} catch (Exception) {
		// On the second iteration you'll get a different exception: "The requested session cannot be accepted..."
		// I think it's because the first attempt to get a session didn't actually finish (?)
	}
	await queueClient.CloseAsync();
}
```

I was able to work around this issue by altering the loop in the example above to check for existing sessions before trying to accept a session:
```
var sessions = await queueClient.GetMessageSessionsAsync();
if (sessions.Any(s => s.SessionId == "mySessionName")) {
	// You can now accept a session successfully without an exception
	//...
}
```
